internal/promqle2etest: Add test for Prometheus counter going through Otel Collector to GCM (and more)

bwplotka opened the pull request:
This is the initial test suite we could use as an acceptance validation and understanding of the latest nuances around various ingestions paths in OSS for GCM.

One immediate use is to test tricky StartTime processing we are fixing in b/389130459

I put it out of CI for now, but we can easily set CI for this, question is do we want to run it on every PR. We do that in prometheus-engine, but you might have a different preference.

See also chained new test cases we work on e.g. https://github.com/bwplotka/otel-ops-col/pull/1

bwplotka commented:
At this point I still can't get otel-gcm pipeline to work, investigating. Somehow it does not send data to GCM.

ridwanmsharif commented on internal/promqle2etest/backend_otelgcm.go:
```diff
@@ -0,0 +1,196 @@
+exporters:
+  googlemanagedprometheus:
+    project: "%s"
+    # TODO(bwplotka): Do we need to disable cumulative_normalization explicitly?
```
We should allow the test to pass in a config imo. We want to test using the `googlemanagedprometheus` exporter but we also want to test that the `otlp` exporter works the same way when used with the `metricstarttime` processor

bwplotka replied:
Yea, totally, I just want to make sure GMP path works too, something timeouts.

bwplotka replied:
Do you have a container image that is built on top of your PR I could use?  https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38594

ridwanmsharif commented on internal/promqle2etest/backend_otelgcm.go:
```diff
@@ -0,0 +1,196 @@
+    # TODO(bwplotka): Do we need to disable cumulative_normalization explicitly?
+    metric:
+      # TODO(bwplotka): Update docs? Change defaults? Those metrics are a bad idea to include (IMO).
+      # We are working in Prometheus OSS to slowly promote more resource attr as labels (if not all one day).
```
Thats an interesting development, have a link handy for these discussions?

bwplotka replied:
Some people still like joining though, so we can keep it by default, just it broke for me (int vs double descriptor issue)

bwplotka commented:
Ok, so generally this works in ~26 secs.

I had to adjust `otel-gcm` labels to expect `instance="..."` label, somehow otel code ignores my relabelling to drop this label, perhaps some magic with `service.instance.id` is happening. 

I also added TODOs for new pipeline with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38594 or so @ridwanmsharif  - feel free to edit this PR/etc 

Closing laptop for today, let me know tomorrow where I can continue with (how to build image from your PR?)

bwplotka commented:
We could also merge if you prefer, this should be now good for review, even with TODOs. Test should pass for anyone with the right GCM SA for their GCP project.

bwplotka commented:
Added more test data for counters. 

Otel collector -> GCM without start time flow seems vary vs Prometheus and Prometheus fork completely:

![image](https://github.com/user-attachments/assets/a6083137-c553-4c10-bf07-f349acaab33f)


However those are a tricky corner cases. Maybe some misconfiguration too? Added current values as "expected for `otel-gcm` to show the issue". Also we might accept this and improve it for OTLP flow @ridwanmsharif is working on. Up to us.

ridwanmsharif commented:
I'll give this a review but +1 to getting this merged and then iterating to figure out why these discrepancies exist.

To answer some questions now though:

> somehow otel code ignores my relabelling to drop this label, perhaps some magic with service.instance.id is happening.

The prometheus receiver sets some values for resource labels here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/prometheusreceiver/internal/prom_to_otlp.go#L47-L72
There is some more logic about when there is no `instance` label found which is probably the culprit: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/d6128eb77788b005f19c31bc5d339548cfa2e64c/receiver/prometheusreceiver/internal/transaction.go#L475-L505

> Maybe some misconfiguration too?

sadly your OTel configuration looks right - you may have just unearthed a bug :)

bwplotka commented:
Thanks and thanks for pre-build image @ridwanmsharif 

With only like ~10m I was able to create a test case for MSTP (metricstarttimeprocessor) + googlemanagedprometheus exporter. I don't want to pollute this base PR if we want to merge it, so I chained my next PR: https://github.com/bwplotka/otel-ops-col/pull/1

As the next steps we could add OTLP case too, but I didn't yet dive into nuances of OTLP GCM and access for it.

Results are pretty good (details in the PR), so I think we could claim googlemanagedprometheus exporter start time adjustment as "broken/deprecated" and simply recommend MSTP with your fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38594

bwplotka commented:
I found why currently recommended flow is "broken" -- the reason is the same as why it's broken for OTLP (the default `prometheusreceiver` broken true reset strategy adjuster -- (b/389130459)).

Luckily we address that with MSTP https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38594#discussion_r2048604963 and this might be the "short" term fix too for everyone.

I adjusted the test commentary to explain all and I merged the MSTP case too to this PR.

cc @ridwanmsharif

dashpole commented on internal/promqle2etest/backend_otelgcm.go:
```diff
@@ -0,0 +1,225 @@
+
+	args := map[string]string{
+		"--config": filepath.Join(f.Dir(), "config.yml"),
+		"--feature-gates=receiver.prometheusreceiver.EnableNativeHistograms": "",
```
did you need to enable the `receiver.prometheusreceiver.RemoveStartTimeAdjustment` feature gate as well?

dashpole commented on internal/promqle2etest/backend_otelgcm.go:
```diff
@@ -0,0 +1,225 @@
+
+# Note that the googlemanagedprometheus exporter block is intentionally blank
+exporters:
+  googlemanagedprometheus:
```
We should consider testing against the OTLP endpoint.  Could be done in a follow-up, or once it is out of preview.
